vertical-pod-autoscaler/pkg/admission-controller/logic/server_test.go

```diff
@@ -374,13 +390,71 @@ func TestGetPatchesForResourceRequest_TwoReplacementResources(t *testing.T) {
 	patches, err := s.getPatchesForPodResourceRequest(podJson, "default")
 	assert.NoError(t, err)
 	// Order of updates for cpu and unobtanium depends on order of iterating a map, both possible results are valid.
-	if assert.Equal(t, len(patches), 3) {
+	if assert.Equal(t, len(patches), 4) {
 		cpuUpdate := addResourceRequestPatch(0, cpu, "1")
 		unobtaniumUpdate := addResourceRequestPatch(0, unobtanium, "2")
 		assert.True(t, eqPatch(patches[0], cpuUpdate) || eqPatch(patches[0], unobtaniumUpdate))
 		assert.True(t, eqPatch(patches[1], cpuUpdate) || eqPatch(patches[1], unobtaniumUpdate))
 		assert.False(t, eqPatch(patches[0], patches[1]))
 		assert.True(t, eqPatch(patches[2], addAnnotationRequest([][]string{{cpu, unobtanium}}, request)) || eqPatch(patches[2], addAnnotationRequest([][]string{{unobtanium, cpu}}, request)))
+		assert.True(t, eqPatch(patches[3], addVpaObservedContainersPatch([]string{})))
+	}
+}
+
+func TestGetPatchesForResourceRequest_VpaObservedContainers(t *testing.T) {
+	tests := []struct {
+		name          string
+		podJson       []byte
+		expectPatches []patchRecord
+	}{
+		{
+			name: "create vpa observed containers annotation",
+			podJson: []byte(
+				`{
+					"spec": {
+						"containers": [
+							{
+								"Name": "test1"
+							},
+							{
+								"Name": "test2"
+							}
+						]
+					}
+				}`),
+			expectPatches: []patchRecord{
+				addVpaObservedContainersPatch([]string{"test1", "test2"}),
+			},
+		},
+		{
+			name: "create vpa observed containers annotation with no containers",
+			podJson: []byte(
+				`{
+					"spec": {
+						"containers": []
+					}
+				}`),
+			expectPatches: []patchRecord{
+				addVpaObservedContainersPatch([]string{}),
+			},
+		},
+	}
+	for _, tc := range tests {
+		t.Run(fmt.Sprintf("test case: %s", tc.name), func(t *testing.T) {
+			fppp := fakePodPreProcessor{}
+			fvpp := fakeVpaPreProcessor{}
+			frp := fakeRecommendationProvider{[]vpa_api_util.ContainerResources{}, vpa_api_util.ContainerToAnnotationsMap{}, "RecomenderName", nil}
+			lc := limitrange.NewNoopLimitsCalculator()
+			s := NewAdmissionServer(&frp, &fppp, &fvpp, lc)
+			patches, _ := s.getPatchesForPodResourceRequest(tc.podJson, "default")
+			if assert.Equal(t, len(tc.expectPatches), len(patches), fmt.Sprintf("got %+v, want %+v", patches, tc.expectPatches)) {
```

Comment:
There is assert.Len (https://godoc.org/github.com/stretchr/testify/assert#Len)

Reply:
Done.
I also added `assert.NoError` for the error returned by `getPatchesForPodResourceRequest`.